runtime/tests/checker/transactions_test.go

```diff
@@ -552,10 +552,72 @@ func TestCheckTransactionRoles(t *testing.T) {
           }
         `)
 
-		errs := RequireCheckerErrors(t, err, 2)
+		errs := RequireCheckerErrors(t, err, 1)
 
 		require.IsType(t, &sema.InvalidTransactionPrepareParameterTypeError{}, errs[0])
-		require.IsType(t, &sema.TypeMismatchError{}, errs[1])
+	})
+
+	t.Run("missing role", func(t *testing.T) {
+		t.Parallel()
+
+		_, err := ParseAndCheck(t, `
+          transaction {
+
+              prepare(first: AuthAccount, second: AuthAccount) {}
+
+              role role1 {
+                  prepare(signer: AuthAccount) {}
+              }
+          }
+        `)
+
+		errs := RequireCheckerErrors(t, err, 1)
+
+		require.IsType(t, &sema.RoleCountMismatchError{}, errs[0])
+	})
+
+	t.Run("too many roles", func(t *testing.T) {
+		t.Parallel()
+
+		_, err := ParseAndCheck(t, `
+          transaction {
+
+              prepare(signer: AuthAccount) {}
+
+              role role1 {
+                  prepare(signer: AuthAccount) {}
+              }
+
+              role role2 {
+                  prepare(signer: AuthAccount) {}
+              }
+          }
+        `)
+
+		errs := RequireCheckerErrors(t, err, 1)
+
+		require.IsType(t, &sema.RoleCountMismatchError{}, errs[0])
+	})
+
+	t.Run("correct number of roles", func(t *testing.T) {
+		t.Parallel()
+
+		_, err := ParseAndCheck(t, `
+          transaction {
+
+              prepare(first: AuthAccount, second: AuthAccount) {}
+
+              role role1 {
+                  prepare(signer: AuthAccount) {}
+              }
+
+              role role2 {
+                  prepare(signer: AuthAccount) {}
+              }
```

Comment:
I still think this feature would benefit from some way to encode in Cadence which of the signers in the outer `prepare` is tied to which role; but that may be beyond the scope of this PR and might just be another discussion point for the FLIP.